File: ios/Classes/PushNativeStore.swift
import Foundation

private extension String {
  func trimmingTrailingSlashes() -> String {
    var s = self
    while s.hasSuffix("/") { s.removeLast() }
    return s
  }
}

struct PushConfigSnapshot {
  let apiBaseUrl: String
  let chatApiPath: String
  let deliveredPathTemplate: String
  let headersJson: String
  let chatTypeDataKey: String
  let chatTypeValue: String

  func buildDeliveredUrl(conversationId: String, messageId: String) -> String {
    var path = deliveredPathTemplate
      .replacingOccurrences(of: "{conversationId}", with: conversationId)
      .replacingOccurrences(of: "{messageId}", with: messageId)
    if !path.hasPrefix("/") {
      path = "/" + path
    }
    var prefix = chatApiPath.trimmingCharacters(in: .whitespacesAndNewlines)
    if !prefix.hasPrefix("/") {
      prefix = "/" + prefix
    }
    prefix = prefix.trimmingTrailingSlashes()
    let base = apiBaseUrl.trimmingCharacters(in: .whitespacesAndNewlines).trimmingTrailingSlashes()
    return "\(base)\(prefix)\(path)"
  }

  func headersMap() -> [String: String] {
    var out: [String: String] = [:]
    guard let data = headersJson.data(using: .utf8),
          let obj = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
      return ["Content-Type": "application/json"]
    }
    for (k, v) in obj {
      out[String(describing: k)] = String(describing: v)
    }
    if out["Content-Type"] == nil {
      out["Content-Type"] = "application/json"
    }
    return out
  }
}

enum PushNativeStore {
  private static let prefix = "hmu_push_"

  private static var defaults: UserDefaults { .standard }

  static func syncPushConfig(_ args: [String: Any]) {
    let d = defaults
    d.set(args["apiBaseUrl"] as? String ?? "", forKey: prefix + "api_base_url")
    d.set(args["chatApiPath"] as? String ?? "/api/v1/chat", forKey: prefix + "chat_api_path")
    d.set(
      args["deliveredPathTemplate"] as? String ?? "/conversations/{conversationId}/messages/{messageId}/delivered",
      forKey: prefix + "delivered_path_template")
    d.set(args["headersJson"] as? String ?? "{}", forKey: prefix + "headers_json")
    d.set(args["chatTypeDataKey"] as? String ?? "type", forKey: prefix + "chat_type_data_key")
    d.set(args["chatTypeValue"] as? String ?? "CHAT_MESSAGE", forKey: prefix + "chat_type_value")
  }

  static func readConfig() -> PushConfigSnapshot? {
    let d = defaults
    let base = (d.string(forKey: prefix + "api_base_url") ?? "").trimmingCharacters(in: .whitespacesAndNewlines)
    if base.isEmpty { return nil }
    return PushConfigSnapshot(
      apiBaseUrl: base,
      chatApiPath: d.string(forKey: prefix + "chat_api_path") ?? "/api/v1/chat",
      deliveredPathTemplate: d.string(forKey: prefix + "delivered_path_template")
        ?? "/conversations/{conversationId}/messages/{messageId}/delivered",
      headersJson: d.string(forKey: prefix + "headers_json") ?? "{}",
      chatTypeDataKey: d.string(forKey: prefix + "chat_type_data_key") ?? "type",
      chatTypeValue: d.string(forKey: prefix + "chat_type_value") ?? "CHAT_MESSAGE"
    )
  }

  static func enqueueAck(conversationId: String, messageId: String) {
    let d = defaults
    let key = prefix + "pending_delivered_ack_queue"
    let raw = d.string(forKey: key) ?? "[]"
    var arr: [[String: Any]] = []
    if let data = raw.data(using: .utf8),
       let parsed = try? JSONSerialization.jsonObject(with: data) as? [[String: Any]] {
      arr = parsed
    }
    arr.append(["conversationId": conversationId, "messageId": messageId])
    if arr.count > 50 {
      arr = Array(arr.suffix(50))
    }
    if let data = try? JSONSerialization.data(withJSONObject: arr),
       let s = String(data: data, encoding: .utf8) {
      d.set(s, forKey: key)
    }
  }

  static func dequeueAll() -> [(String, String)] {
    let d = defaults
    let key = prefix + "pending_delivered_ack_queue"
    let raw = d.string(forKey: key) ?? "[]"
    d.set("[]", forKey: key)
    guard let data = raw.data(using: .utf8),
          let parsed = try? JSONSerialization.jsonObject(with: data) as? [[String: Any]] else {
      return []
    }
    return parsed.compactMap { row in
      let cc = (row["conversationId"] as? String) ?? String(describing: row["conversationId"] ?? "")
      let mm = (row["messageId"] as? String) ?? String(describing: row["messageId"] ?? "")
      guard !cc.isEmpty, !mm.isEmpty else { return nil }
      return (cc, mm)
    }
  }
}

File: android/src/main/kotlin/com/example/health_messenger_ui/PushNativeStore.kt
package com.example.health_messenger_ui

import android.content.Context
import org.json.JSONArray
import org.json.JSONObject

/**
 * Persists auth snapshot and pending delivered-ACK jobs for native FCM handling.
 */
internal object PushNativeStore {
    private const val PREFS = "health_messenger_ui_push"
    private const val KEY_API_BASE = "api_base_url"
    private const val KEY_CHAT_API_PATH = "chat_api_path"
    private const val KEY_DELIVERED_PATH = "delivered_path_template"
    private const val KEY_HEADERS_JSON = "headers_json"
    private const val KEY_TYPE_KEY = "chat_type_data_key"
    private const val KEY_TYPE_VALUE = "chat_type_value"
    private const val KEY_QUEUE = "pending_delivered_ack_queue"

    fun syncPushConfig(context: Context, args: Map<*, *>) {
        val p = context.applicationContext.getSharedPreferences(PREFS, Context.MODE_PRIVATE).edit()
        p.putString(KEY_API_BASE, args["apiBaseUrl"]?.toString() ?: "")
        p.putString(KEY_CHAT_API_PATH, args["chatApiPath"]?.toString() ?: "")
        p.putString(KEY_DELIVERED_PATH, args["deliveredPathTemplate"]?.toString() ?: "")
        p.putString(KEY_HEADERS_JSON, args["headersJson"]?.toString() ?: "{}")
        p.putString(KEY_TYPE_KEY, args["chatTypeDataKey"]?.toString() ?: "type")
        p.putString(KEY_TYPE_VALUE, args["chatTypeValue"]?.toString() ?: "CHAT_MESSAGE")
        p.apply()
    }

    fun readConfig(context: Context): PushConfigSnapshot? {
        val p = context.applicationContext.getSharedPreferences(PREFS, Context.MODE_PRIVATE)
        val base = p.getString(KEY_API_BASE, "") ?: ""
        if (base.isBlank()) return null
        return PushConfigSnapshot(
            apiBaseUrl = base.trimEnd('/'),
            chatApiPath = (p.getString(KEY_CHAT_API_PATH, "/api/v1/chat") ?: "/api/v1/chat").trim(),
            deliveredPathTemplate =
                p.getString(
                    KEY_DELIVERED_PATH,
                    "/conversations/{conversationId}/messages/{messageId}/delivered",
                )
                    ?: "",
            headersJson = p.getString(KEY_HEADERS_JSON, "{}") ?: "{}",
            chatTypeDataKey = p.getString(KEY_TYPE_KEY, "type") ?: "type",
            chatTypeValue = p.getString(KEY_TYPE_VALUE, "CHAT_MESSAGE") ?: "CHAT_MESSAGE",
        )
    }

    fun enqueueAck(context: Context, conversationId: String, messageId: String) {
        val p = context.applicationContext.getSharedPreferences(PREFS, Context.MODE_PRIVATE)
        val raw = p.getString(KEY_QUEUE, "[]") ?: "[]"
        val arr = try {
            JSONArray(raw)
        } catch (_: Exception) {
            JSONArray()
        }
        val obj = JSONObject()
        obj.put("conversationId", conversationId)
        obj.put("messageId", messageId)
        arr.put(obj)
        val capped = JSONArray()
        val start = maxOf(0, arr.length() - 50)
        for (i in start until arr.length()) {
            capped.put(arr.get(i))
        }
        p.edit().putString(KEY_QUEUE, capped.toString()).apply()
    }

    fun dequeueAll(context: Context): List<Pair<String, String>> {
        val p = context.applicationContext.getSharedPreferences(PREFS, Context.MODE_PRIVATE)
        val raw = p.getString(KEY_QUEUE, "[]") ?: "[]"
        p.edit().putString(KEY_QUEUE, "[]").apply()
        val out = mutableListOf<Pair<String, String>>()
        try {
            val arr = JSONArray(raw)
            for (i in 0 until arr.length()) {
                val o = arr.optJSONObject(i) ?: continue
                val c = o.optString("conversationId", "")
                val m = o.optString("messageId", "")
                if (c.isNotBlank() && m.isNotBlank()) {
                    out.add(c to m)
                }
            }
        } catch (_: Exception) {
        }
        return out
    }
}

internal data class PushConfigSnapshot(
    val apiBaseUrl: String,
    val chatApiPath: String,
    val deliveredPathTemplate: String,
    val headersJson: String,
    val chatTypeDataKey: String,
    val chatTypeValue: String,
) {
    fun buildDeliveredUrl(conversationId: String, messageId: String): String {
        var path = deliveredPathTemplate
            .replace("{conversationId}", conversationId)
            .replace("{messageId}", messageId)
        if (!path.startsWith("/")) {
            path = "/$path"
        }
        var prefix = chatApiPath.trim()
        if (!prefix.startsWith("/")) {
            prefix = "/$prefix"
        }
        prefix = prefix.trimEnd('/')
        return "$apiBaseUrl$prefix$path"
    }

    fun headersMap(): Map<String, String> {
        val out = mutableMapOf<String, String>()
        try {
            val o = JSONObject(headersJson)
            val keys = o.keys()
            while (keys.hasNext()) {
                val k = keys.next()
                out[k] = o.optString(k, "")
            }
        } catch (_: Exception) {
        }
        out.putIfAbsent("Content-Type", "application/json")
        return out
    }
}
